### PR TITLE
[TRANSFORMATIONS] Allow NOP-elimination for shapes with dynamic 0th dimension

### PR DESCRIPTION
**Pull request by CuriousPanCake**
NOP-elimination doesn't allow the removal of unnecessary data-movement
operations stacked on top of each other if their shapes are dynamic, even
though certain cases of dynamic shapes can be covered by
NOP-elimination.

Allow NOP-elimination for dynamic shapes if they have only the 0th
dimension dynamic and other static, making this case compatible for
such a fusion.

- Tickets:
  * CVS-158394

Signed-off-by: Andrii Staikov <andrii.staikov@intel.com>

### CODE REVIEW

File: src/common/transformations/src/transformations/common_optimizations/nop_elimination.cpp [hunk 1]
```diff
@@ -156,15 +175,18 @@ static bool eliminate_reshape_v1(const shared_ptr<Node>& node) {
         if (input_node->get_output_target_inputs(0).size() != 1)
             return false;
 
-        auto shape = node->get_output_shape(0);
+        auto shape = node->get_output_partial_shape(0);
 
         // remove interchangeable nodes
-        if (input_node->get_input_partial_shape(0).is_static() && input_node->get_input_shape(0) == shape) {
+        if (input_node->get_input_partial_shape(0).is_static() && input_node->get_input_shape(0) == node->get_output_shape(0)) {
```
Comment (itikhono) [inline src/common/transformations/src/transformations/common_optimizations/nop_elimination.cpp:181]: node->get_output_shape(0) will throw an exception if the output shape has dyn dims

File: src/common/transformations/src/transformations/common_optimizations/nop_elimination.cpp [hunk 2]
```diff
@@ -132,21 +132,40 @@ static bool eliminate_nop(const shared_ptr<Node>& node) {
...
         if (input.get_partial_shape().same_scheme(node->get_output_partial_shape(0)))
             return replace_output_update_name(node->output(0), input);
     }
-
     // check if reshape is not identity op
     if (input.get_partial_shape().is_dynamic() || node->get_output_partial_shape(0).is_dynamic()) {
-        OPENVINO_DEBUG(node, " has dynamic shapes.");
-        return false;
+        if (!only_first_dim_dynamic(input.get_partial_shape()) || !only_first_dim_dynamic(node->get_output_partial_shape(0))) {
+            OPENVINO_DEBUG(node, " has dynamic shapes.");
```
Comment (itikhono) [inline src/common/transformations/src/transformations/common_optimizations/nop_elimination.cpp:162]: please change the debug message

File: src/common/transformations/src/transformations/common_optimizations/nop_elimination.cpp (hunk 1 above)
Comment (CuriousPanCake) [reply src/common/transformations/src/transformations/common_optimizations/nop_elimination.cpp:181]: Fixed.

File: src/common/transformations/src/transformations/common_optimizations/nop_elimination.cpp (hunk 2 above)
Comment (CuriousPanCake) [reply src/common/transformations/src/transformations/common_optimizations/nop_elimination.cpp:162]: Fixed.

**Review by itikhono — APPROVED** [review]

**Pull request merged**